poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+	}
+	copy(buf[3*uint32Size:], pointBuf)
+
+	polyBuf, err := p.pubPoly.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	copy(buf[3*uint32Size+pointLen:], polyBuf)
+
+	// Encode the insurers and secrets array (Based on poly/sharing.go code)
+	bufPos := 3*uint32Size + pointLen + polyLen
+	for i := range p.insurers {
+		pb, err := p.insurers[i].MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		copy(buf[bufPos+i*pointLen:], pb)
+	}
+	bufPos += p.n * pointLen
+
+	for i := range p.secrets {
+		pb, err := p.secrets[i].MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		copy(buf[bufPos+i*secretLen:], pb)
+	}
+	return buf, nil
+}
+
+/* Unmarshals a Promise from a byte buffer
+ *
+ * Arguments
+ *    buf = the buffer containing the Promise
+ *
+ * Returns
+ *   The error status of the unmarshalling (nil if no error)
+ */
+func (p *Promise) UnmarshalBinary(buf []byte) error {
+	pointLen := p.keySuite.PointLen()
+	secretLen := p.shareSuite.SecretLen()
+
+	// Decode n, r, t
+	p.n = int(binary.LittleEndian.Uint32(buf))
+	p.t = int(binary.LittleEndian.Uint32(buf[uint32Size:]))
+	p.r = int(binary.LittleEndian.Uint32(buf[2*uint32Size:]))
+
+	bufPos := 3 * uint32Size
+
+	// Decode pubKey and pubPoly
+	p.pubKey = p.keySuite.Point()
+	if err := p.pubKey.UnmarshalBinary(buf[bufPos : bufPos+pointLen]); err != nil {
+		return err
+	}
+	bufPos += pointLen
+
+	p.pubPoly = poly.PubPoly{}
+	p.pubPoly.Init(p.shareSuite, p.t, nil)
+	polyLen := p.pubPoly.MarshalSize()
+	if err := p.pubPoly.UnmarshalBinary(buf[bufPos : bufPos+polyLen]); err != nil {
+		return err
+	}
+	bufPos += polyLen
+
+	// Decode the insurers and secrets array (Based on poly/sharing.go code)
+	p.insurers = make([]abstract.Point, p.n, p.n)
+	for i := 0; i < p.n; i++ {
+		start := bufPos + i*pointLen
+		end := start + pointLen
+		p.insurers[i] = p.keySuite.Point()
+		if err := p.insurers[i].UnmarshalBinary(buf[start:end]); err != nil {
+			return err
+		}
+	}
+	bufPos += p.n * pointLen
+	p.secrets = make([]abstract.Secret, p.n, p.n)
+	for i := 0; i < p.n; i++ {
+		start := bufPos + i*secretLen
+		end := start + secretLen
+		p.secrets[i] = p.shareSuite.Secret()
+		if err := p.secrets[i].UnmarshalBinary(buf[start:end]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+/* Marshals a Promise struct using an io.Writer
+ *
+ * Arguments
+ *    w = the writer to use for marshalling
+ *
+ * Returns
+ *   The number of bytes written
+ *   The error status of the write (nil if no errors)
+ */
+func (p *Promise) MarshalTo(w io.Writer) (int, error) {
+	buf, err := p.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	return w.Write(buf)
+}
+
+/* Unmarshals a Promise struct using an io.Reader
+ *
+ * Arguments
+ *    r = the reader to use for unmarshalling
+ *
+ * Returns
+ *   The number of bytes read
+ *   The error status of the read (nil if no errors)
+ */
+func (p *Promise) UnmarshalFrom(r io.Reader) (int, error) {
+	// Retrieve p.n and p.t and then initialize p.PubPoly
```

Comment:
This should probably be just a trivial wrapper around UnmarshalBinary(); this much (duplicated?) code shouldn't be needed.